### PR TITLE
feat(inputs.sqlserver): Introduce user specified ID parameter for ADD logins

### LINKED ISSUE #15404: SQL Server AAD Azure Auth method doesn't work if multiple User assigned identies are existing.
### Relevant telegraf.conf

```toml
inputs:
    - sqlserver:
        interval: "30s"
        servers:
          -"Server=dbserver.database.windows.net;Port=1433;database=dbname;hostNameInCertificate=*.database.windows.net;TrustServerCertificate=true;app name=telegraf;log=1;"
        auth_method: "AAD"
        database_type: "AzureSQLDB"
        exclude_query: 
          - "AzureSQLDBSchedulers"
          - "AzureSQLDBRequests"
```

### Logs from Telegraf

```text
[inputs.sqlserver] Error in plugin: error creating AAD token provider for system assigned Azure managed identity : adal: Refresh request failed. Status Code = '400'. Response body: {"error":"invalid_request","error_description":"Multiple user assigned identities exist, please specify the clientId / resourceId of the identity in the token request"}
```

### System info

Telegraf 1.25.0-alpine; Kubernetes 1.30; Azure VM with two MIs

### Docker

_No response_

### Steps to reproduce

1. Create Azure VM and install Kubernetes
2. Assign two identities to the created VM.
3. Create Azure SQL DB
4. Deploy and setup Telegraf

### Expected behavior

The option to select the desired Managed Identity(MI) or to specify its name in the Conf, if more then one MI is assigned to a vm.

At the top of the error log it is called clientId / resourceId. 

### Actual behavior

Authentication to the SQL DB isn't possible.

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by Jan747**
## Summary
If an Azure VM has more than one user assigned identiy, the authentication with the Azure SQL DB does not work. 

Another method to get a token has been introduced for this and a new parameter to select an identy. 

## Checklist

- [x] No AI generated code was used in this PR

## Related issues

resolves #15404

### CODE REVIEW

File: plugins/inputs/sqlserver/sqlserver.go [hunk 1]
```diff
@@ -564,6 +578,7 @@ func init() {
 	inputs.Add("sqlserver", func() telegraf.Input {
 		return &SQLServer{
 			AuthMethod: "connection_string",
+			UserAssignedID: "",
```
Comment (powersj) [inline plugins/inputs/sqlserver/sqlserver.go:581]: This does not need to be here. The default for a string is empty.

```suggestion
```

File: plugins/inputs/sqlserver/sqlserver.go [hunk 2]
```diff
@@ -36,6 +36,8 @@ type SQLServer struct {
 	HealthMetric bool             `toml:"health_metric"`
 	Log          telegraf.Logger  `toml:"-"`
 
+	UserAssignedID string         `toml:"user_assigned_id"`
```
Comment (powersj) [inline plugins/inputs/sqlserver/sqlserver.go:39]: Let's keep the exported variables together. You may need to edit the whole section to ensure the spacing stays consistent.

```suggestion
	HealthMetric   bool             `toml:"health_metric"`
	UserAssignedID string           `toml:"user_assigned_id"`
	Log            telegraf.Logger  `toml:"-"`
```

File: plugins/inputs/sqlserver/sample.conf [hunk 3]
```diff
@@ -19,6 +19,7 @@
   ## Authentication method
   ## valid methods: "connection_string", "AAD"
   # auth_method = "connection_string"
+  # user_assigned_id = "ClientID" # UserAssignedID is the client ID of the user assigned identity that should be used to authenticate to the Azure SQL server.
```
Comment (powersj) [inline plugins/inputs/sqlserver/sample.conf:22]: The config should show the default value and be below a description:

```suggestion
  # User assigned ID of the user assigned identity to authenticate to the SQL
  # server
  # user_assigned_id = "" 
```

Should this actually be called client ID? It is weird to call it user assigned ID, and then refer to it as the client ID.

File: plugins/inputs/sqlserver/README.md [hunk 4]
```diff
@@ -300,6 +300,8 @@ authentication support.
 - Configure "system-assigned managed identity" for Azure resources on the Monitoring VM (the VM that'd connect to the SQL server/database) [using the Azure portal](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/qs-configure-portal-windows-vm).
 - On the database being monitored, create/update a USER with the name of the Monitoring VM as the principal using the below script. This might require allow-listing the client machine's IP address (from where the below SQL script is being run) on the SQL Server resource.
 
+In case of multiple assigened identies on one VM you can use the parameter user_assigned_id to specifiy the client_id.
```
Comment (powersj) [inline plugins/inputs/sqlserver/README.md:303]: Why not call it client_id?

**Review by powersj — COMMENTED** [review]
Thanks for the PR! I appreciate you taking the time to put this up.

I have added some comments. The only big one is what to call the option. You will also need to run `make docs`, or update the sqlserver README.md as well to ensure it has the new config option in the example. make docs does this for you.

**Comment by powersj** [comment]
Also need you to sign the CLA please.

File: plugins/inputs/sqlserver/sample.conf (hunk 3 above)
Comment (Jan747) [reply plugins/inputs/sqlserver/sample.conf:22]: Will rename it to client id

**Comment by Jan747** [comment]
Hello @powersj thank you very much for your comments and tips. 

I improved the PR. Please have another look on it.

**Comment by powersj** [comment]
!signed-cla

**Comment by powersj** [comment]
@Jan747 - Thank you very much for the updates and signing the CLA. I pushed an update to ensure the linter and doc tests pass. Can you confirm you have been able to use your proposed fix and it resolves your issue?

**Comment by powersj** [comment]
@Jan747,

I need to land this early this week so we can make v1.31.0, can you confirm the PR artifacts were able to resolve your issue and you are happy with my updates?

Thanks again!

**Comment by Jan747** [comment]
Hello @powersj  sorry for the late response. I wasn't able to get the version running. Docker image gave error that entrypoint.sh is missing. So I can't say if the error is now fixed or not.

**Comment by macxs** [comment]
I can confirm this build is working. 
@Jan747 @powersj Thank you!

The error message from before is gone now and logs are clean.
```
...
2024-06-13T08:23:20Z I! Starting Telegraf 1.31.0-c549b29c brought to you by InfluxData the makers of InfluxDB
...
2024-06-13T08:23:20Z I! [inputs.sqlserver] Config: database_type: AzureSQLDB , query_version:0 , azuredb: false
2024-06-13T08:23:20Z I! [inputs.sqlserver] Config: Effective Queries: []string{"AzureSQLDBOsWaitstats", "AzureSQLDBResourceStats", "AzureSQLDBWaitStats", "AzureSQLDBDatabaseIO", "AzureSQLDBResourceGovernance", "AzureSQLDBPerformanceCounters"}
```

**Comment by powersj** [comment]
@Jan747 - can you rebase on master and resolve the one conflict? Then we can get this landed.

**Comment by Jan747** [comment]
@macxs thank you very much for testing. 

@powersj confilct is solved. Anything else what I have todo ? And thank you very much again.

**Review by powersj — APPROVED** [review]

**Review by srebhan — APPROVED** [review]
Thanks for your contribution @Jan747!

**Pull request merged**